### ISSUE TITLE
sql: statement diagnostic crashes node with panic

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#63729]
I am able to reliably hit the following panic that brings down a node shortly after activating statement diagnostics through the admin UI:
```
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536  a panic has occurred!
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +runtime error: invalid memory address or nil pointer dereference
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +(1) attached stack trace
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  -- stack trace:
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:482
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | [...repeated from below...]
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +Wraps: (2) while executing: INSERT INTO _(_, _, _, _, _, _, _, _) VALUES ($1, $2, __more6__)
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +Wraps: (3) attached stack trace
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  -- stack trace:
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:482
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | runtime.gopanic
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /usr/local/go/src/runtime/panic.go:969
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError.func1
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:58
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | runtime.gopanic
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /usr/local/go/src/runtime/panic.go:969
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | runtime.panicmem
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /usr/local/go/src/runtime/panic.go:212
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | runtime.sigpanic
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /usr/local/go/src/runtime/signal_unix.go:742
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/kv.(*Txn).Type
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/kv/txn.go:1224
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.GetLeafTxnFinalState
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:271
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*joinReader).generateMeta
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/joinreader.go:759
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.newJoinReader.func1
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/joinreader.go:285
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).moveToTrailingMeta
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:727
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:646
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*joinReader).Next
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/joinreader.go:538
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).DrainHelper
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:642
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Next
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:183
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Next
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:199
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/colflow.(*batchInfoCollector).Next
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/stats.go:86
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).next
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:290
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).nextAdapter
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:315
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/colexecerror.CatchVectorizedRuntimeError
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecerror/error.go:91
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Next
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:321
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.Run
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:175
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:774
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:382
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:359
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).planAndRunPostquery
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1167
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRunCascadesAndChecks
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:1107
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execWithDistSQLEngine
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1058
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:915
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:659
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:122
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  | github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execPortal
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +  |    /go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:209
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +Wraps: (4) runtime error: invalid memory address or nil pointer dereference
[n1,client=‹10.142.0.96:54744›,hostnossl,user=root] 24536 +Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) runtime.errorString
```
CRDB SHA: 7495d19e50d8fd14140220f3ba5c7ad951abafa6

**Comment by yuzefovich** [comment]
Possibly a dup of #62261 (which currently has only been fixed on release-21.1 branch).

**Comment by yuzefovich** [comment]
Yeah, I believe it is a dup of #62261, and I revived the PR to fix it on master.
